src/hotspot/share/cds/aotCodeSource.cpp

```diff
@@ -0,0 +1,987 @@
+#include "logging/log.hpp"
+#include "logging/logStream.hpp"
+#include "memory/metadataFactory.hpp"
+#include "memory/metaspaceClosure.hpp"
+#include "memory/resourceArea.hpp"
+#include "oops/array.hpp"
+#include "oops/objArrayKlass.hpp"
+#include "runtime/arguments.hpp"
+#include "utilities/classpathStream.hpp"
+#include "utilities/formatBuffer.hpp"
+#include "utilities/stringUtils.hpp"
+
+#include <sys/stat.h>
+#include <errno.h>
+
+AOTCodeSourceConfig* AOTCodeSourceConfig::_dumptime_instance = nullptr;
+const AOTCodeSourceConfig* AOTCodeSourceConfig::_runtime_instance = nullptr;
+
+// A CodeSourceStream represents a list of code sources, which can be iterated using
+// start() and has_next().
+class CodeSourceStream {
+protected:
+  GrowableArray<const char*> _array;
+  int _current;
+
+  // Add one path to this stream.
+  void add_one_path(const char* path) {
+    _array.append(path);
+  }
+
+  // Add all paths specified in cp; cp must be from -classpath or -Xbootclasspath/a.
+  void add_paths_in_classpath(const char* cp) {
+    ClasspathStream cp_stream(cp);
+    while (cp_stream.has_next()) {
+      add_one_path(cp_stream.get_next());
+    }
+  }
+
+public:
+  CodeSourceStream() : _array(), _current(0) {}
+
+  void print(outputStream* st) const {
+    const char* sep = "";
+    for (int i = 0; i < _array.length(); i++) {
+      st->print("%s%s", sep, _array.at(i));
+      sep = os::path_separator();
+    }
+  }
+
+  void add(CodeSourceStream& css) {
+    for (css.start(); css.has_next();) {
+      add_one_path(css.get_next());
+    }
+  }
+
+  // Iteration
+  void start() { _current = 0; }
+  bool has_next() const { return _current < _array.length(); }
+  const char* get_next() {
+    return _array.at(_current++);
+  }
+
+  int current() const { return _current; }
+  bool is_empty() const { return _array.length() == 0; }
+};
+
+class BootCpCodeSourceStream : public CodeSourceStream {
+public:
+  BootCpCodeSourceStream() : CodeSourceStream() {
+    // Arguments::get_boot_class_path() contains $JAVA_HOME/lib/modules, but we treat that separately
+    for (const char* bootcp = Arguments::get_boot_class_path(); *bootcp != '\0'; ++bootcp) {
+      if (*bootcp == *os::path_separator()) {
+        ++ bootcp;
+        add_paths_in_classpath(bootcp);
+        break;
+      }
+    }
+  }
+};
+
+class AppCpCodeSourceStream : public CodeSourceStream {
+public:
+  AppCpCodeSourceStream() : CodeSourceStream() {
+    const char* appcp = Arguments::get_appclasspath();
+    if (strcmp(appcp, ".") == 0) {
+      appcp = "";
+    }
+    add_paths_in_classpath(appcp);
+  }
+};
+
+class ModulePathCodeSourceStream : public CodeSourceStream {
+  bool _has_non_jar_modules;
+public:
+  ModulePathCodeSourceStream();
+  bool has_non_jar_modules() { return _has_non_jar_modules; }
+};
+
+// AllCodeSourceStreams is used to iterate over all the code sources that
+// are available to the application from -Xbootclasspath, -classpath and --module-path
```

Comment:
Consider adding this comment:

// When creating an AOT cache, we store the contents from AllCodeSourceStreams
// into an array of AOTCodeSources. See AOTCodeSourceConfig::dumptime_init_helper().
// When loading the AOT cache in a production run, we compare the contents of the
// stored AOTCodeSources against the current AllCodeSourceStreams to determine whether
// the AOT cache is compatible with the current JVM. See AOTCodeSourceConfig::validate().

Reply:
Added the comment.